### ISSUE TITLE
인간정보자산최적배치연구 - #4, 데이터학습(머신러닝)

### ISSUE THREAD

**Issue by KHJ818** [KHJ818/humintplusgeoai#4]
import pandas as pd
import numpy as np
import glob
import os
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from sklearn.utils import resample
import joblib

# === 경로 설정 ===
FINAL_TOP300_DIR = r"파일경로"
FILTERED_DIR = r"파일경로"
MODEL_DIR = r"폴더경로"
os.makedirs(MODEL_DIR, exist_ok=True)

# === 저장 모델 경로 ===
RF_MODEL_PATH = os.path.join(MODEL_DIR, 'rf_final_model.pkl')
XGB_MODEL_PATH = os.path.join(MODEL_DIR, 'xgb_final_model.pkl')
LGBM_MODEL_PATH = os.path.join(MODEL_DIR, 'lgbm_final_model.pkl')

# === 데이터 병합 ===
dfs = []

for top300_file in glob.glob(os.path.join(FINAL_TOP300_DIR, "*_최종학습300_분산.csv")):
    df_top300 = pd.read_csv(top300_file)
    base_name = os.path.basename(top300_file).replace("_최종학습300_분산.csv", "")
    filtered_file = os.path.join(FILTERED_DIR, f"{base_name}.csv")

    if not os.path.exists(filtered_file):
        print(f"⚠️ 필터링 파일 없음: {filtered_file}")
        continue

    df_filtered = pd.read_csv(filtered_file)

    # 정답 레이블 1
    df_top300['label'] = 1
    # 비정답 레이블 0 (정답과 동일 수 샘플링)
    df_filtered = df_filtered[~df_filtered[['grid_x', 'grid_y']].apply(tuple, axis=1).isin(
        df_top300[['grid_x', 'grid_y']].apply(tuple, axis=1)
    )]
    df_filtered_sampled = resample(df_filtered, n_samples=len(df_top300), random_state=42)
    df_filtered_sampled['label'] = 0

    combined = pd.concat([df_top300, df_filtered_sampled], ignore_index=True)
    dfs.append(combined)

    print(f"✅ 병합 완료: {base_name} | 정답: {len(df_top300)} | 비정답 샘플링: {len(df_filtered_sampled)}")

# === 전체 데이터 병합 ===
if not dfs:
    raise ValueError("❌ 병합할 학습 데이터가 없습니다!")

df_all = pd.concat(dfs, ignore_index=True)
print(f"🧠 전체 학습용 샘플 수: {len(df_all)}")

# === Feature 설정 ===
features = ['grid_x', 'grid_y', '고도', 'distance(m)', 'l2_name', 'road_type']
X = df_all[features]
y = df_all['label']  # 0 또는 1

# === 데이터 분할 ===
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# === 1️⃣ 랜덤포레스트 학습 ===
print("\n🌲 랜덤포레스트 학습 시작...")
rf_model = RandomForestClassifier(n_estimators=200, max_depth=10, random_state=42)
rf_model.fit(X_train, y_train)

y_pred_rf = rf_model.predict(X_test)
print("\n📊 랜덤포레스트 평가 결과:\n")
print(classification_report(y_test, y_pred_rf, digits=4))
joblib.dump(rf_model, RF_MODEL_PATH)
print(f"\n✅ 랜덤포레스트 모델 저장 완료: {RF_MODEL_PATH}")

# === 2️⃣ XGBoost 학습 ===
print("\n🚀 XGBoost 학습 시작...")
xgb_model = XGBClassifier(n_estimators=200, max_depth=6, random_state=42)
xgb_model.fit(X_train, y_train)

y_pred_xgb = xgb_model.predict(X_test)
print("\n📊 XGBoost 평가 결과:\n")
print(classification_report(y_test, y_pred_xgb, digits=4))
joblib.dump(xgb_model, XGB_MODEL_PATH)
print(f"\n✅ XGBoost 모델 저장 완료: {XGB_MODEL_PATH}")

# === 3️⃣ LightGBM 학습 ===
print("\n💡 LightGBM 학습 시작...")
lgbm_model = LGBMClassifier(n_estimators=200, max_depth=6, random_state=42)
lgbm_model.fit(X_train, y_train)

y_pred_lgbm = lgbm_model.predict(X_test)
print("\n📊 LightGBM 평가 결과:\n")
print(classification_report(y_test, y_pred_lgbm, digits=4))
joblib.dump(lgbm_model, LGBM_MODEL_PATH)
print(f"\n✅ LightGBM 모델 저장 완료: {LGBM_MODEL_PATH}")

print("\n🎉 모든 모델 학습 및 저장 완료! (정답+비정답)")
